app/react/Markdown/components/EntityData.tsx

```diff
@@ -0,0 +1,150 @@
+
+const extractMetadataProperty = ({ formattedEntity, property, newNameGeneration }: Options) => {
+  const propertyData = formattedEntity.metadata.find(
+    (p: any) => p.name === safeName(property, newNameGeneration)
+  );
+  return showByType(propertyData, false);
+};
+
+const extractRootLabel = ({ property, template: _template }: Options) => {
+  const template = ensure<IImmutable<ClientTemplateSchema>>(_template);
+  const term =
+    template
+      .get('commonProperties')
+      ?.find(p => p?.get('name') === property)
+      .get('label') || '';
+
+  const context = property === 'title' ? template.get('_id') : 'System';
+  return <Translate context={context}>{term}</Translate>;
+};
+
+const extractMetadataLabel = ({ formattedEntity, property, newNameGeneration }: Options) => {
+  const propertyData = getPropertyData({ formattedEntity, property, newNameGeneration });
+  return <Translate context={propertyData.translateContext}>{propertyData.label}</Translate>;
+};
+
+const logError = (err: any, propValueOf?: string, propLabelOf?: string) => {
+  /* eslint-disable no-console */
+  console.log('Error on EntityData: ');
+  console.log('value-of: ', propValueOf, '; label-of: ', propLabelOf);
+  console.log(err);
+  /* eslint-enable no-console */
+};
+
+const getProperty = (
+  propValueOf?: EntityDataProps['value-of'],
+  propLabelOf?: EntityDataProps['label-of']
+) => {
+  if (propValueOf && propLabelOf) {
+    throw new Error('Can\'t provide both "value-of" and "label-of".');
+  }
+
+  const property = propValueOf || propLabelOf;
+
+  if (!property) {
+    throw new Error('"value-of" or "label-of" must be provided.');
+  }
+
+  return property;
+};
+
+const getMethod = (propValueOf: string | undefined, isRootProperty: boolean) => {
+  let method: Function = () => {};
+
+  if (propValueOf && isRootProperty) {
+    method = extractRootProperty;
+  }
+
+  if (propValueOf && !isRootProperty) {
+    method = extractMetadataProperty;
+  }
+
+  if (!propValueOf && isRootProperty) {
+    method = extractRootLabel;
+  }
+
+  if (!propValueOf && !isRootProperty) {
+    method = extractMetadataLabel;
+  }
+
+  return method;
+};
+
+const prepareData = (
+  propValueOf?: EntityDataProps['value-of'],
+  propLabelOf?: EntityDataProps['label-of']
+) => {
+  const property = getProperty(propValueOf, propLabelOf);
+  const isRootProperty = rootProperties.includes(property);
+  const method = getMethod(propValueOf, isRootProperty);
+
+  return { method, property };
+};
+
+const EntityData = ({
+  entity,
+  templates,
+  thesauri,
+  'value-of': propValueOf,
+  'label-of': propLabelOf,
+  settings,
+}: ComponentProps) => {
+  const newNameGeneration = settings.collection.get('newNameGeneration') || false;
+  const formattedEntity = formatter.prepareMetadata(entity.toJS(), templates, thesauri);
+  let output = <></>;
+
+  try {
+    const { property, method } = prepareData(propValueOf, propLabelOf);
```

Comment:
i think names here are too generic, especially the method name, which seems to do more than 1 thing, maybe that is why is hard to name ?

Reply:
Renamed method to renderMethod. It can do four different things:
- extractRootProperty
- extractMetadataProperty
- extractRootLabel
- extractMetadataLabel

I think it reads better now, let me now.

Reply:
my main concern was the method name, i do not think prepareData says anything, actually i think you can move what that method does into the render method and everything is clearer.

Reply:
Yes, I agree!  Tell that to the complexity analyzer and eslint function length.  If they agree, I am more than happy to change that! :)